stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,262 @@
+
+        Args:
+            neo4j_graphdb: (py2neo.Graph) the Neo4J Graph Database object
+            nodes (list): A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n_size (int): The number of neighbouring nodes to expand at each depth of the walk. Sampling of
+            n (int, default 1): Number of walks per node id.
+            neighbours with replacement is always used regardless of the node degree and number of neighbours
+            requested.
+            seed (int, optional): Random number generator seed; default is None
+
+        Returns:
+            A list of lists such that each list element is a sequence of ids corresponding to a BFW.
+        """
+        self._check_sizes(n_size)
+        self._check_common_parameters(nodes, n, len(n_size), seed)
+
+        bfsQuery = self.breadth_first_search_query_builder(n_size)
+
+        result_records = neo4j_graphdb.run(bfsQuery, {"head_nodes": nodes})
+        walks = pd.DataFrame(result_records)[0][0]
+
+        return walks
+
+
+@experimental(reason="the class is not fully tested")
+class Neo4JDirectedBreadthFirstNeighbors(GraphWalk):
+    """
+    Breadth First sampler that generates the composite of a number of sampled paths from a starting node.
+    It can be used to extract a random sub-graph starting from a set of initial nodes from the Neo4J graph database.
+    """
+
+    def __init__(self, graph, graph_schema=None, seed=None):
+        super().__init__(graph, graph_schema, seed)
+        if not graph.is_directed():
+            self._raise_error("Graph must be directed")
+
+    def directed_breadth_first_search_query_builder(self, in_samples, out_samples):
+        """
+        Build the Cypher query to perform a sampled directed breadth-first search walk from root nodes.
+
+        Args:
+            in_samples (int): The number of in-directed nodes to sample with replacement at each depth of the walk.
+            out_samples (int): The number of out-directed nodes to sample with replacement at each depth of the walk.
+
+        Returns:
+            Cypher query to retrieve all node ids of the sampled subgraph.
+        """
+
+        query = "\n".join(
+            (
+                # find the head_nodes
+                "OPTIONAL MATCH (head_node) WHERE ID(head_node) in {head_nodes}",
+                # put the current level of nodes into a list of list
+                "WITH [apoc.coll.flatten(collect(head_node))] AS cur_depth_list,",
+                # add the list of node ids into subgraph
+                "[apoc.coll.flatten(collect(id(head_node)))] AS subgraph",
+            )
+        )
+
+        for in_num, out_num in zip(in_samples, out_samples):
+            next_hop_query = "\n".join(
+                (
+                    # unpack cur_depth_list which is a list of list of nodes, into single list of nodes
+                    "UNWIND cur_depth_list AS cur_node_list",
```

Comment:
Can we write these big queries as one big string instead? like:
```python
next_hop_query = """
UNWIND cur_depth_list AS cur_node_list
CALL apoc.cypher.run('
...
"""
```
Although I guess that probably means we can't put comments on specific lines. I feel like it's a bit difficult to read at the moment.

Reply:
Also, do you think it's possible to take some parts of the query out into functions so we can reuse them in the two walker classes?

Reply:
Right, I think we could definitely turn this into a long string `f"""UNWIND {some_variables}"""` with parameters. 
I will look seriously into whether we could redesign the query for reusing purposes!

Reply:
Using multiline strings sounds like a great idea: the `"\n".join(...)` trick with literals is clever, but it results in a lot of extra syntax with all the quotes and commas and, as Kevin said, it makes it hard to read.

> I will look seriously into whether we could redesign the query for reusing purposes!

I think this is good as a "long term" thing, but for now it'd be great to focus on getting the queries into a state that can be more easily read (that is, multiline strings). It's relatively easy to factor out common code, after the code is landed, but it's hard to land the code if we can't easily review/understand it.